### ISSUE TITLE
Faraday Timeout Error uploading files to BigQuery

### ISSUE THREAD

**Issue by gramos74** [GoogleCloudPlatform/google-cloud-ruby#570]
I have been loading a lot of files to BigQuery and from time to time I receive a

Faraday::TimeoutError: execution expired

There is a way to change the timeout configuration that Faraday is using ?

**Comment by quartzmo** [comment]
@gramos74 I tried the following, which did NOT appear to work for me:

```ruby
conn = Faraday.default_connection
conn.options[:timeout] = 1
data = bigquery.query sql # request takes > 1s
```

Then, I searched through google-api-client 0.8.6 for any mention of timeout, and [found this example of how to pass Faraday options](https://github.com/google/google-api-ruby-client/blob/v0.8.6/spec/google/api_client_spec.rb#L63).

It appears that such options would need to be exposed as a parameter and passed [here in the gcloud BigQuery connection](https://github.com/GoogleCloudPlatform/gcloud-ruby/blob/v0.6.1/lib/gcloud/bigquery/connection.rb#L37-L38)?

This is where I concluded my investigation, but perhaps @blowmage knows of another solution or approach?

**Comment by blowmage** [comment]
I can think of a couple ways to set a timeout. First, you could create a custom Faraday middleware that sets the timeout. Second, you could monkey punch your way through gcloud-ruby's use of the Google API Client, and Google API Client's use of Faraday and set the timeout on the Faraday connection options:

```ruby
# set new faraday timeout of 2 minutes
bigquery.connection.instance_variable_get("@client").connection.options.timeout = 120
```

The problem with both these approaches is that they will stop working as Google API Client moves from Faraday to Hurley, and as gcloud-ruby moves from Google API Client to GRPC.

The design approach taken by gcloud-ruby so far has been to hide the connection details as much as possible, so we don't know if or how to allow timeout to be configured in the public API. This is something we will have to give more thought on.

@gramos74 Ideally, how and where would you like to configure a timeout when using gcloud-ruby?

**Comment by gramos74** [comment]
Hi,

The how is not really clear, specially after reading your comment. I agree that is a good design keeping the connection details hidden.

Where ? In our app, right now, the most critical point where this exception is causing 'problems' is in the upload of the files to BigQuery. Usually fails just after send the last part of the file or checking the status job of the load. So, in both cases is in the LoadJob. If we considere that the LoadJob does a lot of http requests, the dependency in the connection behaviour of this specific job is high. 

This could justify 'breaking' the design?  To be honest, I don't know.

**Comment by quartzmo** [comment]
Personally, I prefer designs that "leak" access to their dependencies. I am completely open to changing gcloud-ruby to do so in this case.

**Comment by quartzmo** [comment]
@gramos74 While we discuss potential changes, can you confirm that a Faraday configuration change via this workaround solves your problem?

```ruby
# set new faraday timeout of 2 minutes
bigquery.connection.instance_variable_get("@client").connection.options.timeout = 120
```

**Comment by gramos74** [comment]
Yes, I tested it locally and seems to work fine. But tomorrow I will push it to production and there I will check it with huge files. I will let you know how it goes.

**Comment by gramos74** [comment]
Hi, Sorry for the delay but I was too much busy the last two days.

Let me put first some numbers. The last 5 days I uploaded more than 10000 files to GBQ inserting more than 1600000000 rows.

After changing the timeout value for Faraday as you suggested, I only received 5 Faraday::TimeoutError exceptions, so I would say that the change improved the process definitely.

Hope it can help in order to take a decision.

**Comment by quartzmo** [comment]
@gramos74 Your usage certainly makes a strong case for the value of being able to set the timeout. I am strongly in favor of changing the design of the library to support such "under-the-hood" access to connection details for users who need it.

Within the next week I will make an attempt to do so in a PR. This will be a speculative proposal-- merely a concrete example for further discussion.

**Comment by gramos74** [comment]
With the workaround proposed above the upload worked fine until today. Today I have some huge files (2.1 Gb) to be uploaded and the upload process failed: "integer 2245126166 too big to convert to `int' "

Checking the backtrace I saw the problem is in the transport layer and it is related to the size of the string (file).

/usr/local/rvm/rubies/ruby-2.1.1/lib/ruby/2.1.0/openssl/buffering.rb:326:in `syswrite'

In the gcloud table source code [lib/gcloud/bigquery/table.rb] I saw that the method load_resumable supports the chunk_size option, but it is not supported in the load method.

If I add support for this option in the load method I can load those huge files ( already tested with a successful result ) with a chunk_size of 1Gb.

Would be possible to add the support for chunk_size in the load_method ?

**Comment by blowmage** [comment]
@gramos74 I think this is a regression. You should be able to set `chunk_size` on the `load` method. We will get that restored ASAP.

Also, when a `chunk_size` is not set, the current behavior is to leave it undefined, which means Faraday will send the entire file in one chunk, but that seems to not work for files larger than 2GB. So I'm thinking we should set a default chunk_size in those cases. What do you think is a sensible default for that chunk size? 250MB? 500MB? 1GB? 2GB?

**Comment by gramos74** [comment]
Well, seems that is not working in the ssl transport, so in this case is not a problem of Faraday :p

You can find more info here: https://bugs.ruby-lang.org/issues/10633

I would say that a default size of 1Gb is ok.

**Comment by gramos74** [comment]
BTW... I already have a fork of the project because I need to fix it asap in production. So I can prepare a pull request once it is ready if you agree.

**Comment by blowmage** [comment]
@gramos74 Go ahead and submit the PR. We are also actively working on it.

**Comment by quartzmo** [comment]
This was fixed in PR #702 and refactored for reuse in Storage in #703.

**Comment by blowmage** [comment]
@gramos74 Thoughts on #788?

**Comment by quartzmo** [comment]
Closing, as I believe this was addressed by #788.
